Allow for arbitrary QSPI alt sizes

kyle-cypress opened the pull request:
### Description
The QSPI spec allows alt to be any size that is a multiple of the number of data lines. For example, Micron's N25Q128A uses only a single alt cycle for all read modes (1, 2, or 4 bits depending on
how many data lines are in use).
<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@ARMmbed/team-cypress
<!--
    Optional
    Request additional reviewers with @username
-->

## Release notes

### Reason for change
The QSPI interface permits the alt (aka mode) bits to be any multiple of the bus width. But the values in qspi_alt_size_t only support multiples of bytes. While #11295 masks the issue, this is insufficient to support the memory devices on all targets.

### Analysis of effects
No functional impact on existing code. QSPI driver clients may, but are not required to, utilize the greater flexibility in alt sizes.
Allows parts which require non-byte-aligned alt bits to continue working after #11295 is fixed, which prevents QSPIFBlockDevice from functioning correctly with memory devices which require a non-empty alt value for one or more of the operations which QSPIFBlockDevice performs.

### Migration path
No change is required to existing code. Compatibility macros are provided for the old qspi_alt_size_t values.
New code is recommended to explicitly specify the required number of alt bits rather than using the 
QSPI_CFG_ALT_SIZE_* macros.
<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

SeppoTakalo commented on components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.cpp:
```diff
@@ -247,7 +247,7 @@ int QSPIFBlockDevice::init()
     // Configure  BUS Mode to 1_1_1 for all commands other than Read
     _qspi_configure_format(QSPI_CFG_BUS_SINGLE, QSPI_CFG_BUS_SINGLE, QSPI_CFG_ADDR_SIZE_24, QSPI_CFG_BUS_SINGLE,
-                           QSPI_CFG_ALT_SIZE_8, QSPI_CFG_BUS_SINGLE, 0);
+                           0, QSPI_CFG_BUS_SINGLE, 0);
```
I don't understand this change. Documentation says:
```
*  @param alt_size Size in bits used by alt phase(Valid values are QSPI_CFG_ALT_SIZE_8, QSPI_CFG_ALT_SIZE_16, QSPI_CFG_ALT_SIZE_24, QSPI_CFG_ALT_SIZE_32)
```

So why zero instead of one of the valid values?

SeppoTakalo replied:
Sorry, you changed it to:
```
*  @param alt_size Size in bits used by alt phase (must be a multiple of the number of bus lines indicated in alt_width)
```

But I still don't get it, why zero is multiple of bus lines?

kyle-cypress replied:
Prior to this change, QSPI_CFG_ALT_SIZE_8 seems to be chosen as a somewhat arbitrary placeholder (the code has no particular knowledge that 8 alt bits are actually what is required) which then winds up being ignored because QSPIFBlockDevice never passes anything other than -1 for alt. 
Setting 0 (no alt bits) seems like a more direct way of expressing what QSPIFBlockDevice is doing. However, I'm okay reverting this to 8 for now if you'd prefer given the more significant rework of alt bit handling in #11531 .

SeppoTakalo replied:
No need to revert. Explanation is enough. Zero is OK to express "No alt bits"

LMESTM commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -116,47 +116,70 @@ void qspi_prepare_command(const qspi_command_t *command, OSPI_RegularCmdTypeDef
+            
+        /* command->AlternateBytesSize needs to be shifted by OCTOSPI_CCR_ABSIZE_Pos */
+        // 0b00 = 1 byte, 0b01 = 2 bytes, 0b10 = 3 bytes, 0b11 = 4 bytes
+        st_command->AlternateBytesSize = ((rounded_size - 1) << OCTOSPI_CCR_ABSIZE_Pos) & OCTOSPI_CCR_ABSIZE_Msk;
```
Please rework this part to use a switch case and assign st_command->AlternateBytesSize only with definitions of HAL_OSPI_ALTERNATE_BYTES_8_BITS, HAL_OSPI_ALTERNATE_BYTES_16_BITS, etc.
The actual values assignment (0, 01, 02 ) and its position in the register might change from a target to another  ...

kyle-cypress replied:
Updated to introduce get_alt_bytes_size for this purpose

LMESTM commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -249,10 +278,39 @@ void qspi_prepare_command(const qspi_command_t *command, QSPI_CommandTypeDef *st
-        st_command->AlternateBytesSize = (command->alt.size << QUADSPI_CCR_ABSIZE_Pos) & QUADSPI_CCR_ABSIZE_Msk;
-        st_command->AlternateBytesSize = command->alt.size;
+        // 0b00 = 1 byte, 0b01 = 2 bytes, 0b10 = 3 bytes, 0b11 = 4 bytes
+        st_command->AlternateBytesSize = ((rounded_size - 1) << QUADSPI_CCR_ABSIZE_Pos) & QUADSPI_CCR_ABSIZE_Msk;
```
Same remark as above, please use switch case and assign st_command->AlternateBytesSize only with definitions of HAL_QSPI_ALTERNATE_BYTES_8_BITS, HAL_QSPI_ALTERNATE_BYTES_16_BITS, etc ...

I know this was not done like this before but tests were all okay, so we lived with it. Now that we need to rework to cope with a new condition, better to clean it up.

LMESTM commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -230,15 +255,19 @@ void qspi_prepare_command(const qspi_command_t *command, QSPI_CommandTypeDef *st
         st_command->AddressSize = (command->address.size << QUADSPI_CCR_ADSIZE_Pos) & QUADSPI_CCR_ADSIZE_Msk;
```
As part of rework, better need to also assign AddressSize with switch case and assignment from one of definitions: HAL_OSPI_ADDRESS_8_BITS, HAL_OSPI_ADDRESS_16_BITS, etc...

kyle-cypress replied:
I don't think this change belongs in this pull request, which right now doesn't do anything related to AddressSize. I'd rather keep this PR focused on the driver interface change, with the minimum target updates required to keep things working.

LMESTM commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -116,47 +116,70 @@ void qspi_prepare_command(const qspi_command_t *command, OSPI_RegularCmdTypeDef
         }
+
+        // Round up to nearest byte - unused parts of byte act as dummy cycles
+        uint32_t rounded_size = ((command->alt.size - 1) >> 3) + 1;
```
Could you add some comments to explain what is being done here ? 
command->alt.size is a number of bits right ? before the change it was one of 8 / 16 / 24 / 32  bits, right ?
what are the possible values now ? Any from 1 to 32 and we're rounding this value to what ? nearest _upper_ byte or nearest byte ?

from what I see 
1 bit is being rounded up to 1 byte
7 bits is being rounded up to 1 byte
18 bits is being rounded to 3 bytes
31 bits to 4 bytes 
so I guess that would be the nearest up byte ...

kyle-cypress replied:
Yes, rounded_size is up to the nearest byte. I renamed the variable to make this more clear.

LMESTM commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -249,10 +278,39 @@ void qspi_prepare_command(const qspi_command_t *command, QSPI_CommandTypeDef *st
+            st_command->DummyCycles -= integrated_dummy_cycles;
+
+            // Align alt value to the end of the most significant byte
+            st_command->AlternateBytes = command->alt.value << leftover_bits;
```
Isn't there a check or a mask that the AlternateBytes value width actually fits into the AlternateBytesSize ?
e.g. make sure that if 3 bits size were selected, then AlternateBytes isn't bigger than 3 bits ?

kyle-cypress replied:
There isn't an existing check that I can see, and this PR does not change that. I'm also not sure how one would write such a check. You could approximate by checking the position of the last non-zero bit, but that doesn't give you any guarantee because there could be trailing zeros as part of the required mode bits. Unfortunately I think this just has to trust that the caller set consistent values. 
(additionally, as above, any change for this should not be in this PR).

LMESTM reviewed (changes requested):
hi
as this change reworks part of the driver, I would like to take the opportunity to ask for using the HAL_OSPI and HAL_QSPI definitions wherever possible. Also I'm asking a few questions for my understanding of this change  ...

Commit:
Update STM driver changes for clarity

- Use a switch statement rather than shifting and masking to compute
  the AlternateBytes value.
- Rename rounded_size to alt_bytes to clarify its purpose.

LMESTM reviewed (approved):
LGTM for target_stm related changes - thanks !

0xc0170 reviewed (changes requested):
@kyle-cypress Please add Release notes section, following the template : 

https://os.mbed.com/docs/mbed-os/v5.14/contributing/workflow.html#pull-request-template-functional-changes

Commit:
Restore old qspi_alt_size_t values as compatibility macros

0xc0170 reviewed (changes requested):
Something is not correct after the latest rebase or? I can see here only macros addition and STM target change. Please review @kyle-cypress
